rids/network_capture.py

```diff
@@ -14,74 +14,103 @@
+        '-f', f'ip and src ip == {self._host_ip}',
+        '-Tfields',
+        # The order of the following fields determines the ordering in output
+        '-e', 'frame.time',
+        '-e', 'ip.dst',
+        '-l',
+    ])
 
-def detect_tls_events(input_stream):
-    unusual_tls_traffic = {}  # map[int]dict stream_id -> connection_details
+    for line in iter(proc.stdout.readline, b''):
+      values = line.split('\t')
+      ip_info = observations.IpPacket(
+        timestamp=values[0],
+        remote_ip=ipaddress.ip_address(values[1]))
+      yield ip_info
 
-    logging.basicConfig(level=logging.DEBUG,
-                        format='%(asctime)s %(levelname)-8s %(message)s',
-                        datefmt='%m-%d %H:%M',
-                        filename='/var/maldetector.log',
-                        filemode='a')
 
-    for line in iter(input_stream, b''):
-        values = line.split('\t')
-        if len(values) >= 8:
-            stream_id = int(values[1])
-            if int(values[2]) == 1:  # client hello
-                sni_and_port = (values[7], int(values[6]))
-                if sni_and_port not in allowed_sni_port:
-                    unusual_tls_traffic[stream_id] = {
-                        'client_ts': values[0],
-                        'server_ip': values[5],
-                        'server_port': values[6],
-                        'server_name': values[7],
-                        'ja3': values[8],
-                        'ja3_full': values[9],
-                    }
-            elif int(values[2]) == 2:  # server hello
-                port = int(values[4])
-                if port == 443:
-                    continue
-                connection_details = unusual_tls_traffic.get(stream_id, None)
-                if not connection_details:
-                    continue
-                connection_details['ja3s'] = values[8]
-                connection_details['ja3s_full'] = values[9]
-                logging.info('connection_details %s', connection_details)
-                del unusual_tls_traffic[stream_id]
+class TlsConnectionMonitor:
+  """Continuously scans the network traffic for TLS handshakes.
+  The scan() function is an iterator that produces a tuple of
+    (client_hello, server_hello)
+  where server_hello may be `None` if only the client hello has been seen, and
+  will include both when the server_hello is found.  Each is a dict that has
+  `remote_ip`, `remote_port`, `sni` (if present), `ja3`, `ja3s` as properties.
+  """
+  def __init__(self, host_ip):
+    self._tls_streams = {}
+    self._host_ip = host_ip
 
-            print(line.rstrip())
+  def monitor(self):
+    """Generator for observations of unusual TLS traffic."""
+    proc = _start_tshark([
+        '-f', 'tcp and not (src port 443 or dst port 443)',
+        '-Y', (f'(tls.handshake.type == 1 and ip.src == {self._host_ip})' +
+               f'or (tls.handshake.type == 2 and ip.dst == {self._host_ip})'),
+        '-Tfields',
```

Comment:
Nice to have, but can be in another PR:
There are a few JSON output options that could make the parsing code clearer. See `-T rk|json|jsonraw`
https://www.wireshark.org/docs/man-pages/tshark.html

Reply:
yeah, I remember looking into the json output and preferring the tsv default for some reason, maybe the duplicate keys and noncompliant JSON output, but it seems those problems have been fixed in more recent versions.  I don't think I looked at `ek` or `jsonraw` either.  I have read the man page a few times.

Reply:
I'll follow up on this in another PR, if json doesn't leave out any fields that we need then I think the added clarity is good.  It will likely be at the cost of some I/O and CPU overhead, I plan on measuring that soon too.

It helps that the tshark fields and the processing of its output are now in the same module, but it does get confusing when some fields can be missing, as in our TLS connection monitor.

Reply:
Don't worry about the CPU until it actually becomes an issue. There's a lot more going on per packet that takes a lot more cpu.

Reply:
I suspect the I/O overhead will bite us before the CPU overhead, and yeah we probably don't have to worry about it, but I think we should still measure it so we can know when/whether it is approaching becoming an issue.  Especially since our processing load is directly tied to (typically highly variable) network load, and may be in a multi-user environment where available resources are tight.  I'm not immediately worried about it, either, and I'm not trying to prematurely optimize, but features like JSON-formatted intermediate representation are, as you say, a nice-to-have.